torchaudio/prototype/datasets/bucketize_sampler.py

```diff
@@ -0,0 +1,141 @@
+            num_buckets (int): The number of buckets.
+            min_len (int): The lower bound of the evenly spaced length intervals to determine bucket width.
+            max_len (int): The upper bound of the evenly spaced length intervals to determine bucket width.
+
+        Returns:
+            (dict[int, Tensor]): A dictionary in which the key is the bucket index, the value is
+                the Tensor of corresponding sample indices.
+        """
+        buckets = {}
+
+        boundaries = [min_len - 1]
+        interval = (max_len - min_len) // num_buckets
+        for i in range(1, num_buckets):
+            boundaries.append(min_len + i * interval)
+        boundaries.append(max_len + 1)
+        bucket_ids = torch.bucketize(torch.tensor(lengths), torch.tensor(boundaries))
+        for i in indices:
+            bucket_id = bucket_ids[self.index2iter[i]]
+            if bucket_id in buckets:
+                buckets[bucket_id].append(i)
+            else:
+                buckets[bucket_id] = [i]
+        for k in buckets:
+            buckets[k] = torch.as_tensor(buckets[k], dtype=torch.int)
+            if self.shuffle:
+                buckets[k] = buckets[k][torch.randperm(buckets[k].size(0))]
+        return buckets
+
+    def __iter__(self) -> Iterator[List[int]]:
+        iter_list = []
+        total_len = 0
+        batch = []
+        if self.max_token_count is not None:
+            for k in self.buckets.keys():
+                for i in range(self.buckets[k].size(0)):
+                    index = self.buckets[k][i]
+                    if len(batch) == 0 and self.lengths[self.index2iter[int(index)]] > self.max_token_count:
+                        raise ValueError(
+                            f"The length of the sample is {self.lengths[self.index2iter[int(index)]]}, "
+                            f"which is greater than ``max_token_count``: {self.max_token_count}."
+                            "Make sure you set the correct ``max_len`` and ``max_token_count``."
+                        )
+                    elif total_len + self.lengths[self.index2iter[int(index)]] <= self.max_token_count:
+                        batch.append(int(index))
+                        total_len += self.lengths[self.index2iter[int(index)]]
+                    else:
+                        iter_list.append(batch)
+                        batch = []
+                        total_len = 0
+
+            if len(batch) > 0:
+                iter_list.append(batch)
+        else:
+            for k in self.buckets.keys():
+                for i in range(self.buckets[k].size(0)):
+                    index = self.buckets[k][i]
+                    if total_len == self.batch_size:
+                        iter_list.append(batch)
+                        batch = [index]
+                        total_len = 1
+                    else:
+                        batch.append(index)
+                        total_len += 1
```

Comment:
also, i think this might drop the last batch — should perhaps use the same logic as above to add the last partial batch to `iter_list` and update `__len__` to return the ceiling of `len(self.lengths) / self.batch_size`

Reply:
Got it. I will add ``drop_last`` argument and a note saying this argument is only valid when ``batch_size`` is given.